### ISSUE TITLE
Fix false positives for JS variables in value-keyword-case

### ISSUE THREAD

**Issue by jsg2021** [stylelint/stylelint#5185]
> Clearly describe the bug

stylelint is checking and flagging/fixing javascript as if it were css.

> Which rule, if any, is the bug related to?

all? `value-keyword-case` in this example.

> What code is needed to reproduce the bug?

some file named with `.jsx` extension

demo.jsx:
```jsx
function Foo(props) {
	const myBackground = '....';
	const style = {
		...props.style,

		backgroundImage: myBackground ? myBackground : 'url(...)',
	};

	return <div {...props} style={style} />;
}
```

> What stylelint configuration is needed to reproduce the bug?

e.g.

```json
{
  "extends": [
    "stylelint-config-standard"
  ]
}
```

> Which version of stylelint are you using?

`13.11.0`

> How are you running stylelint: CLI, PostCSS plugin, Node.js API?

VS code extension and cli... my cli ignores JS files for now. But when I save in vs code, it breaks if I allow its save-action.

> Does the bug relate to non-standard syntax (e.g. SCSS, Less etc.)?

no

> What did you expect to happen?

No validation/fixing on js. CSS in tag template literals is fine to check because that IS css. But JS properties should not be validated.

> What actually happened (e.g. what warnings or errors did you get)?

CSS errors are flagged. (and fixes attempted)

```shell
demo.jsx
 4:2   ✖  Expected indentation of 4 spaces               indentation       
 6:2   ✖  Expected indentation of 4 spaces               indentation       
 6:19  ✖  Expected "myBackground" to be "mybackground"   value-keyword-case
 6:34  ✖  Expected "myBackground" to be "mybackground"   value-keyword-case
 7:2   ✖  Expected indentation of 2 spaces               indentation
```

**Comment by jeddy3** [comment]
@jsg2021 Thanks for the report and for using the template.

I can reproduce this using [the demo](https://stylelint.io/demo#N4Igxg9gJgpiBcIBmBXAdmALgSwmgBAGIQQAUADgE4TkDOAlPsADprOaRq2b4C2AngCEAhmADWAc2roo+ALz4A5ADpVyxQG5W7Tt3zd+AGxjym2zOzVUatZQeMAabWwuYARqMnS0UAJK9hCRh4fAADARFxKQgZfAB+PiFPaNiQxRRKQ1I1ekVQpxcAXy02F0oYTAyCAB4obAA3JitqOkL9TCMYOWB7GDaAegA+EsKQBxBafjRMYQAPBHBaWgBabDRlgCtaMfA8JGwJBZYCfGYQShRjWjOQ4-x705B64UMUGGWxGH4AdwhKKGWYGEtBgN0ehgg3xglDOrHuhVYo0KQA).

It's likely a problem with [the parser](https://github.com/stylelint/postcss-css-in-js) stylelint uses [to support CSS-in-JS](https://stylelint.io/about/syntaxes). I've added this issue to https://github.com/stylelint/stylelint/issues/4574, which is a growing list of the problems with the syntax that we need help to fix.

Please consider [contributing to the parser](https://stylelint.io/developer-guide/syntaxes) if you have time.

---

> VS code extension and cli... my cli ignores JS files for now. But when I save in vs code, it breaks if I allow its save-action.

You can either:
- [configure stylelint to ignore JS files](https://stylelint.io/user-guide/ignore-code#files-entirely) - the VS Code extension will respect this
- [configure the VS Code extension to only validate CSS files](https://github.com/stylelint/vscode-stylelint#stylelintvalidate), e.g. `validate: ["css"]`.

---

> No validation/fixing on js. CSS in tag template literals is fine to check because that IS css. But JS properties should not be validated.

The majority of [the built-in rules](https://stylelint.io/user-guide/rules/list) are applicable to CSS-in-JS object notation, especially the [possible errors](https://stylelint.io/user-guide/rules/list#possible-errors) and [feature limiting](https://stylelint.io/user-guide/rules/list#limit-language-features) rules. Even a number of [stylistic](https://stylelint.io/user-guide/rules/list#stylistic-issues) rules are applicable to strings within object notation, e.g. the following:

```js
const Button = styled("button", {
  width: "Min( 100VW ,  .1%)"
}
```

Will be fixed to:

```js
const Button = styled("button", {
  width: "min(100vw,  0.1%)"
}
```

**Comment by jsg2021** [comment]
Thanks!

> > No validation/fixing on js. CSS in tag template literals is fine to check because that IS css. But JS properties should not be validated.
> 
> The majority of [the built-in rules](https://stylelint.io/user-guide/rules/list) are applicable to CSS-in-JS object notation, especially the [possible errors](https://stylelint.io/user-guide/rules/list#possible-errors) and [feature limiting](https://stylelint.io/user-guide/rules/list#limit-language-features) rules. Even a number of [stylistic](https://stylelint.io/user-guide/rules/list#stylistic-issues) rules are applicable to strings within object notation, e.g. the following:
> 
> ```js
> const Button = styled("button", {
>   width: "Min( 100VW ,  .1%)"
> }
> ```
> 
> Will be fixed to:
> 
> ```js
> const Button = styled("button", {
>   width: "min(100vw,  0.1%)"
> }
> ```

ah, that's a good point. My statement was too broad. I was specifically thinking about identifiers, not string literals.

**Comment by jsg2021** [comment]
I've added this to my code config to prevent js/css lint conflicting fix loops:
```
"editor.codeActionsOnSave": {
    "source.fixAll": true,
    "source.fixAll.stylelint": false
  },
```

**Comment by jsg2021** [comment]
Did some digging.
https://github.com/stylelint/stylelint/blob/16ae9e4537975d1d755e430a813a74ddfb681cde/lib/rules/value-keyword-case/index.js#L65

At this point, the value in the example is a JS expression, and the `decl.parent.type` is `object`... parsing the js expression as a CSS value. I would guess that it should ignore non-string-literal values in the expression?

**Comment by jeddy3** [comment]
Thanks for digging.

> I would guess that it should ignore non-string-literal values in the expression?

The CSS-in-JS syntax should probably substitute JS expressions for dummy values in the PostCSS AST.

For example:

```js
backgroundImage: myBackground ? myBackground : 'url(...)',
```

Becomes:

```json
{
  "type": "root",
  "nodes": [
    {
      "type": "decl",
      "prop": "background-image",
      "value": "$dummyvalue"
    }
  ],
}
```

There's [experimental work](https://github.com/stylelint/postcss-css-in-js/pull/46) looking into doing this for interpolation in template literals.

**Comment by ahuglajbclajep** [comment]
There is a same problem with `function-name-case`.
`stylelint-disable` comments do not work for these.

**Comment by Alecell** [comment]
Same here, I've been trying to ignore the `${}` pattern but without success.

There isn't a work around on this?

**Comment by rbong** [comment]
> The CSS-in-JS syntax should probably substitute JS expressions for dummy values in the PostCSS AST.

Is that the best approach? It just seems like `postcss-value-parser` isn't meant to parse CSS-in-JS values. Maybe a new parser could help traverse CSS-in-JS values more accurately? I know that it would be a lot of work to develop a new parser and update rules, but all of the information you need to distinguish Javascript variables from strings is in the node, it's only discarded when the declaration value is read as a plain string and parsed.

**Comment by kachkaev** [comment]
I think I’m having the same issue in JSX:

```tsx
const MyComponent = () => {
  const minWidth = 42;
  return <div style={{ minWidth }}>hello</div>;
}
```

This shows: `Expected "minWidth" to be "minwidth" (value-keyword-case) stylelint(value-keyword-case)`.

We already have `.stylelintignore` with the following contents:

```
## file extensions
*.*
LICENSE
!*.css
!*.less


## shared between linters
...


## copied from .gitignore
...
```

(this approach to .stylelintignore` file structure is described in https://github.com/prettier/prettier/issues/6280#issuecomment-509568953)

All works in CLI, but because of a [bug in the VSCode extension](https://github.com/stylelint/vscode-stylelint/issues/155), the false positive still shows up in the editor. It is possible to set `"stylelint.validate": ["css", "less"]` in VSCode, but that only works for user preferences, so cannot be shared with the team via `[project]/.vscode/settings.json`.

Interestingly, replacing `style={{ minWidth }}` with `blah={{ minWidth }}` removes the error. Is there any workaround that would help ignore `value-keyword-case` by saying ‘don’t look at jsx → style prop’? 🤔

**Comment by jeddy3** [comment]
Closing [as stylistic rules are frozen](https://github.com/stylelint/stylelint/issues/5674). The community is welcome to migrate the rule [to a plugin](https://stylelint.io/developer-guide/plugins) and fix this bug.
